ClassCastException: NodeImpl -> NodeContent

## Gentics Mesh Version, operating system, or hardware.
- v1.1.1

## Operating System
- Linux / Docker-Container

## JVM

openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.3+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.3+7, mixed mode)

## Problem

I was upgrading an Application from 0.22.x to 1.1.1. While doing so, I was also updating the API to V2 and therefore migrating the GraphQL Queries.

After posting a query to the graphql Endpoint, the request hung up and I checked the logs to find this in the logs:

```
mesh_1           | Exception in thread "vert.x-eventloop-thread-0" io.reactivex.exceptions.OnErrorNotImplementedException: The exception was not handled due to missing onError handler in the subscribe() method call. Further reading: https://github.com/ReactiveX/RxJava/wiki/Error-Handling | java.lang.ClassCastException: class com.gentics.mesh.core.data.node.impl.NodeImpl cannot be cast to class com.gentics.mesh.core.data.node.NodeContent (com.gentics.mesh.core.data.node.impl.NodeImpl and com.gentics.mesh.core.data.node.NodeContent are in unnamed module of loader 'app')
mesh_1           |      at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
mesh_1           |      at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
mesh_1           |      at io.reactivex.internal.operators.maybe.MaybeCallbackObserver.onError(MaybeCallbackObserver.java:82)
mesh_1           |      at io.reactivex.internal.operators.maybe.MaybeDelayWithCompletable$DelayWithMainObserver.onError(MaybeDelayWithCompletable.java:107)
mesh_1           |      at io.vertx.reactivex.impl.AsyncResultMaybe.lambda$subscribeActual$0(AsyncResultMaybe.java:60)
mesh_1           |      at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:330)
mesh_1           |      at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
mesh_1           |      at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
mesh_1           |      at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
mesh_1           |      at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
mesh_1           |      at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:333)
mesh_1           |      at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
mesh_1           |      at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
mesh_1           |      at java.base/java.lang.Thread.run(Unknown Source)
mesh_1           | Caused by: java.lang.ClassCastException: class com.gentics.mesh.core.data.node.impl.NodeImpl cannot be cast to class com.gentics.mesh.core.data.node.NodeContent (com.gentics.mesh.core.data.node.impl.NodeImpl and com.gentics.mesh.core.data.node.NodeContent are in unnamed module of loader 'app')
mesh_1           |      at com.gentics.mesh.graphql.type.NodeTypeProvider.lambda$null$3(NodeTypeProvider.java:206)
mesh_1           |      at graphql.execution.ExecutionStrategy.resolveTypeForInterface(ExecutionStrategy.java:741)
mesh_1           |      at graphql.execution.ExecutionStrategy.resolveType(ExecutionStrategy.java:713)
mesh_1           |      at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:412)
mesh_1           |      at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:363)
mesh_1           |      at graphql.execution.ExecutionStrategy.lambda$resolveFieldWithInfo$0(ExecutionStrategy.java:201)
mesh_1           |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(Unknown Source)
mesh_1           |      at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(Unknown Source)
mesh_1           |      at java.base/java.util.concurrent.CompletableFuture.thenApply(Unknown Source)
mesh_1           |      at graphql.execution.ExecutionStrategy.resolveFieldWithInfo(ExecutionStrategy.java:200)
mesh_1           |      at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:70)
mesh_1           |      at graphql.execution.ExecutionStrategy.completeValueForObject(ExecutionStrategy.java:633)
mesh_1           |      at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:413)
mesh_1           |      at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:363)
mesh_1           |      at graphql.execution.ExecutionStrategy.lambda$resolveFieldWithInfo$0(ExecutionStrategy.java:201)
mesh_1           |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(Unknown Source)
mesh_1           |      at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(Unknown Source)
mesh_1           |      at java.base/java.util.concurrent.CompletableFuture.thenApply(Unknown Source)
mesh_1           |      at graphql.execution.ExecutionStrategy.resolveFieldWithInfo(ExecutionStrategy.java:200)
mesh_1           |      at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:70)
mesh_1           |      at graphql.execution.Execution.executeOperation(Execution.java:158)
mesh_1           |      at graphql.execution.Execution.execute(Execution.java:100)
mesh_1           |      at graphql.GraphQL.execute(GraphQL.java:554)
mesh_1           |      at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:496)
mesh_1           |      at graphql.GraphQL.executeAsync(GraphQL.java:470)
mesh_1           |      at graphql.GraphQL.execute(GraphQL.java:401)
mesh_1           |      at com.gentics.mesh.graphql.GraphQLHandler.lambda$handleQuery$0(GraphQLHandler.java:73)
mesh_1           |      at io.vertx.reactivex.core.Vertx$3.handle(Vertx.java:643)
mesh_1           |      at io.vertx.reactivex.core.Vertx$3.handle(Vertx.java:641)
mesh_1           |      at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
mesh_1           |      at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
mesh_1           |      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
mesh_1           |      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
mesh_1           |      ... 2 more
```

This application contains sensible data, therefore I provide the data privately.
It contains the GraphQL, the **original** (before 1.1.1 update) `graphdb` and `binaryFiles`/`uploads` data.

## Reproducer

Post the appended graphql query within the provided dump as admin.

## Expected behaviour and actual behaviour

No errors or a proper Error response which cancels the Request

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -23,6 +23,10 @@ include::content/docs/variables.adoc-include[]
 
 * The support for the embedded Elasticsearch will be dropped in the future. It is highly recommended to link:{{< relref "elasticsearch.asciidoc" >}}#_dedicated_elasticsearch[setup Elasticsearch as a dedicated service].
 
+[[Unreleased]]
+
+icon:check[] GraphQL: Fixed fetching node references from users. link:https://github.com/gentics/mesh/issues/393[#393] link:https://github.com/gentics/mesh/issues/934[#934]
+
 [[v1.2.0]]
 
 == 1.2.0 (20.11.2019)

--- a/core/src/test/java/com/gentics/mesh/core/graphql/GraphQLEndpointTest.java
+++ b/core/src/test/java/com/gentics/mesh/core/graphql/GraphQLEndpointTest.java
@@ -78,6 +78,8 @@ import com.gentics.mesh.core.rest.schema.impl.NumberFieldSchemaImpl;
 import com.gentics.mesh.core.rest.schema.impl.SchemaReferenceImpl;
 import com.gentics.mesh.core.rest.schema.impl.StringFieldSchemaImpl;
 import com.gentics.mesh.core.rest.user.NodeReference;
+import com.gentics.mesh.core.rest.user.UserResponse;
+import com.gentics.mesh.core.rest.user.UserUpdateRequest;
 import com.gentics.mesh.parameter.impl.PublishParametersImpl;
 import com.gentics.mesh.parameter.impl.VersioningParametersImpl;
 import com.gentics.mesh.rest.client.MeshRestClient;
@@ -179,7 +181,8 @@ public class GraphQLEndpointTest extends AbstractMeshTest {
 			Arrays.asList("filtering/roles", true, "draft"),
 			Arrays.asList("node/breadcrumb-root", true, "draft"),
 			Arrays.asList("node/versionslist", true, "draft"),
-			Arrays.asList("permissions", true, "draft")
+			Arrays.asList("permissions", true, "draft"),
+			Arrays.asList("user-node-reference", true, "draft")
 		)
 		.flatMap(testCase -> IntStream.rangeClosed(1, CURRENT_API_VERSION).mapToObj(version -> {
 			// Make sure all testData entries have five parts.
@@ -456,6 +459,11 @@ public class GraphQLEndpointTest extends AbstractMeshTest {
 		nodeUpdateRequest.getFields().put("nodeList", FieldUtil.createNodeListField(NODE_WITH_LINKS_UUID));
 		nodeUpdateRequest.getFields().put("slug", FieldUtil.createStringField("node-with-reference-de"));
 		call(() -> client.updateNode(PROJECT_NAME, NODE_WITH_NODE_REF_UUID, nodeUpdateRequest));
+
+		// Add node reference to user
+		UserResponse user = call(() -> client.me());
+		NodeResponse newsNode = call(() -> client.findNodeByUuid(PROJECT_NAME, NEWS_UUID));
+		call(() -> client.updateUser(user.getUuid(), new UserUpdateRequest().setNodeReference(newsNode)));
 
 		// Now execute the query and assert it
 		GraphQLResponse response = call(

--- a/core/src/test/resources/graphql/user-node-reference
+++ b/core/src/test/resources/graphql/user-node-reference
@@ -1,0 +1,9 @@
+{
+    me {
+        nodeReference {
+            # [$.data.me.nodeReference.uuid=<is-uuid>]
+            uuid
+        }
+    }
+}
+# [$.errors=<is-undefined>]

--- a/verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/UserTypeProvider.java
+++ b/verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/UserTypeProvider.java
@@ -1,16 +1,5 @@
 package com.gentics.mesh.graphql.type;
 
-import com.gentics.mesh.core.data.User;
-import com.gentics.mesh.core.data.node.Node;
-import com.gentics.mesh.etc.config.MeshOptions;
-import com.gentics.mesh.graphql.context.GraphQLContext;
-import graphql.schema.GraphQLObjectType;
-import graphql.schema.GraphQLObjectType.Builder;
-import graphql.schema.GraphQLTypeReference;
-
-import javax.inject.Inject;
-import javax.inject.Singleton;
-
 import static com.gentics.mesh.core.data.relationship.GraphPermission.READ_PERM;
 import static com.gentics.mesh.core.data.relationship.GraphPermission.READ_PUBLISHED_PERM;
 import static com.gentics.mesh.graphql.type.GroupTypeProvider.GROUP_PAGE_TYPE_NAME;
@@ -19,6 +8,22 @@ import static graphql.Scalars.GraphQLBoolean;
 import static graphql.Scalars.GraphQLString;
 import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
 import static graphql.schema.GraphQLObjectType.newObject;
+
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.gentics.mesh.core.data.NodeGraphFieldContainer;
+import com.gentics.mesh.core.data.User;
+import com.gentics.mesh.core.data.node.Node;
+import com.gentics.mesh.core.data.node.NodeContent;
+import com.gentics.mesh.etc.config.MeshOptions;
+import com.gentics.mesh.graphql.context.GraphQLContext;
+
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLObjectType.Builder;
+import graphql.schema.GraphQLTypeReference;
 
 @Singleton
 public class UserTypeProvider extends AbstractTypeProvider {
@@ -108,7 +113,11 @@ public class UserTypeProvider extends AbstractTypeProvider {
 					// TODO throw error - We can't traverse across projects
 					return null;
 				}
-				return gc.requiresPerm(node, READ_PERM, READ_PUBLISHED_PERM);
+
+				node = gc.requiresPerm(node, READ_PERM, READ_PUBLISHED_PERM);
+				List<String> languageTags = getLanguageArgument(env);
+				NodeGraphFieldContainer container = node.findVersion(gc, languageTags);
+				return new NodeContent(node, container, languageTags);
 			}));
 
 		return root.build();

